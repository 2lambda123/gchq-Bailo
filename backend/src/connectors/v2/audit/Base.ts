import { Request } from 'express'

import { AccessRequestDoc } from '../../../models/v2/AccessRequest.js'
import { ModelDoc } from '../../../models/v2/Model.js'
import { ReleaseDoc } from '../../../models/v2/Release.js'
import { ModelSearchResult } from '../../../routes/v2/model/getModelsSearch.js'
import { BailoError } from '../../../types/v2/error.js'

const AuditKind = {
  Create: 'Create',
  View: 'View',
  Update: 'Update',
  Delete: 'Delete',
  Search: 'Search',
}
export type AuditKindKeys = (typeof AuditKind)[keyof typeof AuditKind]

export const AuditInfo = {
  CreateModel: { typeId: 'CreateModel', description: 'Model Created', auditKind: AuditKind.Create },
  ViewModel: { typeId: 'ViewModel', description: 'Model Viewed', auditKind: AuditKind.View },
  UpdateModel: { typeId: 'UpdateModel', description: 'Model Updated', auditKind: AuditKind.Update },
  SearchModels: { typeId: 'SearchModels', description: 'Model Searched', auditKind: AuditKind.Search },

  CreateRelease: { typeId: 'CreateRelease', description: 'Release Created', auditKind: AuditKind.Create },
  ViewRelease: { typeId: 'ViewRelease', description: 'Release Viewed', auditKind: AuditKind.View },
  UpdateRelease: { typeId: 'UpdateRelease', description: 'Release Updated', auditKind: AuditKind.Update },
  DeleteRelease: { typeId: 'UpdateRelease', description: 'Release Deleted', auditKind: AuditKind.Update },
  SearchReleases: { typeId: 'SearchReleases', description: 'Release Searched', auditKind: AuditKind.Search },

  CreateAccessRequest: {
    typeId: 'CreateAccessRequest',
    description: 'Access Request Created',
    auditKind: AuditKind.Create,
  },
  ViewAccessRequest: { typeId: 'ViewAccessRequest', description: 'Access Request Viewed', auditKind: AuditKind.View },
  UpdateAccessRequest: {
    typeId: 'UpdateAccess Request',
    description: 'Access Request Updated',
    auditKind: AuditKind.Update,
  },
  DeleteAccessRequest: {
    typeId: 'UpdateAccessRequest',
    description: 'Access Request Deleted',
    auditKind: AuditKind.Update,
  },
  SearchAccessRequests: {
    typeId: 'SearchAccessRequests',
    description: 'Access Request Searched',
    auditKind: AuditKind.Search,
  },
}
export type AuditInfoKeys = (typeof AuditInfo)[keyof typeof AuditInfo]

export abstract class BaseAuditConnector {
  abstract onCreateModel(req: Request, model: ModelDoc)
  abstract onViewModel(req: Request, model: ModelDoc)
  abstract onSearchModel(req: Request, models: ModelSearchResult[])

  abstract onCreateRelease(req: Request, release: ReleaseDoc)
  abstract onViewRelease(req: Request, release: ReleaseDoc)
  abstract onUpdateRelease(req: Request, release: ReleaseDoc)
  abstract onDeleteRelease(req: Request, modelId: string, semver: string)
  abstract onSearchReleases(req: Request, releases: ReleaseDoc[])

  abstract onCreateAccessRequest(req: Request, accessRequest: AccessRequestDoc)
  abstract onViewAccessRequest(req: Request, accessRequest: AccessRequestDoc)
  abstract onUpdateAccessRequest(req: Request, accessRequest: AccessRequestDoc)
  abstract onDeleteAccessRequest(req: Request, accessRequestId: string)
  abstract onSearchAccessRequests(req: Request, accessRequests: AccessRequestDoc[])

  abstract onError(req: Request, error: BailoError)
}
