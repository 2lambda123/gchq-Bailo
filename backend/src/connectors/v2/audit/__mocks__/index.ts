import { vi } from 'vitest'

const audit = {
  onCreateModel: vi.fn(),
  onViewModel: vi.fn(),
  onUpdateModel: vi.fn(),
  onSearchModel: vi.fn(),

  onCreateModelCard: vi.fn(),
  onViewModelCard: vi.fn(),
  onViewModelCardRevisions: vi.fn(),
  onUpdateModelCard: vi.fn(),

  onViewFiles: vi.fn(),
  onDeleteFile: vi.fn(),
  onCreateFile: vi.fn(),

  onCreateRelease: vi.fn(),
  onViewRelease: vi.fn(),
  onUpdateRelease: vi.fn(),
  onDeleteRelease: vi.fn(),
  onSearchReleases: vi.fn(),

  onCreateAccessRequest: vi.fn(),
  onViewAccessRequest: vi.fn(),
  onUpdateAccessRequest: vi.fn(),
  onDeleteAccessRequest: vi.fn(),
  onSearchAccessRequests: vi.fn(),

  onSearchReviews: vi.fn(),
  onCreateReviewResponse: vi.fn(),

  onCreateSchema: vi.fn(),
  onViewSchema: vi.fn(),
  onDeleteSchema: vi.fn(),
  onSearchSchemas: vi.fn(),

  onViewModelImages: vi.fn(),

  onError: vi.fn(),
}
export default audit
