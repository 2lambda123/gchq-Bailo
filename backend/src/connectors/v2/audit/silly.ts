import { Request } from 'express'

import { AccessRequestDoc } from '../../../models/v2/AccessRequest.js'
import { FileInterface, FileInterfaceDoc } from '../../../models/v2/File.js'
import { ModelCardInterface, ModelDoc, ModelInterface } from '../../../models/v2/Model.js'
import { ReleaseDoc } from '../../../models/v2/Release.js'
import { ReviewInterface } from '../../../models/v2/Review.js'
import { SchemaInterface } from '../../../models/v2/Schema.js'
import { ModelSearchResult } from '../../../routes/v2/model/getModelsSearch.js'
import log from '../../../services/v2/log.js'
import { BailoError } from '../../../types/v2/error.js'
import { AuditInfo, AuditInfoKeys, BaseAuditConnector } from './Base.js'

interface Outcome {
  Success: boolean
  Description: string
  Permitted?: boolean
}

export class SillyAuditConnector extends BaseAuditConnector {
  constructor() {
    super()
  }

  /**
   * Model Events
   */

  onCreateModel(req: Request, model: ModelDoc) {
    this.checkEventType(AuditInfo.CreateModel, req)
    const event = this.generateEvent(req, { id: model.id })
    log.info(event, 'Logging Event')
  }

  onViewModel(req: Request, model: ModelDoc) {
    this.checkEventType(AuditInfo.ViewModel, req)
    const event = this.generateEvent(req, { name: model.name })
    log.info(event, 'Logging Event')
  }

  onUpdateModel(req: Request, model: ModelDoc) {
    this.checkEventType(AuditInfo.UpdateModel, req)
    const event = this.generateEvent(req, { id: model.id })
    log.info(event, 'Logging Event')
  }

  onSearchModel(req: Request, models: ModelSearchResult[]) {
    this.checkEventType(AuditInfo.SearchModels, req)
    const event = this.generateEvent(req, {
      url: req.originalUrl,
      results: models.map((model) => model.id),
    })

    log.info(event, 'Logging Event')
  }

  onCreateModelCard(req: Request, modelId: string, modelCard: ModelCardInterface) {
    this.checkEventType(AuditInfo.CreateModelCard, req)
    const event = this.generateEvent(req, { modelId, version: modelCard.version })
    log.info(event, 'Logging Event')
  }

  onViewModelCard(req: Request, modelId: string, modelCard: ModelCardInterface) {
    this.checkEventType(AuditInfo.ViewModelCard, req)
    const event = this.generateEvent(req, { modelId, version: modelCard.version })
    log.info(event, 'Logging Event')
  }

  onUpdateModelCard(req: Request, modelId: string, modelCard: ModelCardInterface) {
    this.checkEventType(AuditInfo.UpdateModelCard, req)
    const event = this.generateEvent(req, { modelId, version: modelCard.version })
    log.info(event, 'Logging Event')
  }

  onViewModelCardRevisions(req: Request, modelCards: ModelCardInterface[]) {
    this.checkEventType(AuditInfo.ViewModelCardRevisions, req)
    const event = this.generateEvent(req, { url: req.originalUrl, results: modelCards.map((model) => model.version) })
    log.info(event, 'Logging Event')
  }

  onCreateFile(req: Request, file: FileInterfaceDoc) {
    this.checkEventType(AuditInfo.CreateFile, req)
    const event = this.generateEvent(req, { id: file._id, modelId: file.modelId })
    log.info(event, 'Logging Event')
  }

  onViewFiles(req: Request, modelId: string, files: FileInterface[]) {
    this.checkEventType(AuditInfo.ViewFiles, req)
    const event = this.generateEvent(req, { modelId, results: files.map((file) => file.name) })
    log.info(event, 'Logging Event')
  }

  onDeleteFile(req: Request, modelId: string, fileId: string) {
    this.checkEventType(AuditInfo.DeleteFile, req)
    const event = this.generateEvent(req, { modelId, fileId })
    log.info(event, 'Logging Event')
  }

  /**
   * Release Events
   */

  onCreateRelease(req: Request, release: ReleaseDoc) {
    this.checkEventType(AuditInfo.CreateRelease, req)
    const event = this.generateEvent(req, { modelId: release.modelId, semver: release.semver })
    log.info(event, 'Logging Event')
  }

  onViewRelease(req: Request, release: ReleaseDoc) {
    this.checkEventType(AuditInfo.ViewRelease, req)
    const event = this.generateEvent(req, { modelId: release.modelId, semver: release.semver })
    log.info(event, 'Logging Event')
  }

  onUpdateRelease(req: Request, release: ReleaseDoc) {
    this.checkEventType(AuditInfo.UpdateRelease, req)
    const event = this.generateEvent(req, { modelId: release.modelId, semver: release.semver })
    log.info(event, 'Logging Event')
  }

  onDeleteRelease(req: Request, modelId: string, semver: string) {
    this.checkEventType(AuditInfo.DeleteRelease, req)
    const event = this.generateEvent(req, { modelId, semver })
    log.info(event, 'Logging Event')
  }

  onSearchReleases(req: Request, releases: ReleaseDoc[]) {
    this.checkEventType(AuditInfo.SearchReleases, req)
    const event = this.generateEvent(req, {
      url: req.originalUrl,
      results: releases.map((release) => ({ modelId: release.modelId, semver: release.semver })),
    })
    log.info(event, 'Logging Event')
  }

  onCreateAccessRequest(req: Request, accessRequest: AccessRequestDoc) {
    this.checkEventType(AuditInfo.CreateAccessRequest, req)
    const event = this.generateEvent(req, { id: accessRequest.id })
    log.info(event, 'Logging Event')
  }

  onViewAccessRequest(req: Request, accessRequest: AccessRequestDoc) {
    this.checkEventType(AuditInfo.ViewAccessRequest, req)
    const event = this.generateEvent(req, { id: accessRequest.id })
    log.info(event, 'Logging Event')
  }

  onUpdateAccessRequest(req: Request, accessRequest: AccessRequestDoc) {
    this.checkEventType(AuditInfo.UpdateAccessRequest, req)
    const event = this.generateEvent(req, { id: accessRequest.id })
    log.info(event, 'Logging Event')
  }

  onDeleteAccessRequest(req: Request, accessRequestId: string) {
    this.checkEventType(AuditInfo.DeleteAccessRequest, req)
    const event = this.generateEvent(req, { accessRequestId })
    log.info(event, 'Logging Event')
  }

  onSearchAccessRequests(req: Request, accessRequests: AccessRequestDoc[]) {
    this.checkEventType(AuditInfo.SearchAccessRequests, req)
    const event = this.generateEvent(req, {
      url: req.originalUrl,
      results: accessRequests.map((accessRequest) => ({
        id: accessRequest.id,
      })),
    })
    log.info(event, 'Logging Event')
  }

  onSearchReviews(req: Request, reviews: (ReviewInterface & { model: ModelInterface })[]) {
    this.checkEventType(AuditInfo.SearchReviews, req)
    const event = this.generateEvent(req, {
      url: req.originalUrl,
      results: reviews.map((review) => ({
        modelId: review.modelId,
        ...(review.semver && { semver: review.semver }),
        ...(review.accessRequestId && { semver: review.accessRequestId }),
      })),
    })
    log.info(event, 'Logging Event')
  }

  onCreateReviewResponse(req: Request, review: ReviewInterface) {
    this.checkEventType(AuditInfo.CreateAccessRequest, req)
    const event = this.generateEvent(req, {
      modelId: review.modelId,
      ...(review.semver && { semver: review.semver }),
      ...(review.accessRequestId && { semver: review.accessRequestId }),
    })
    log.info(event, 'Logging Event')
  }

  onError(req: Request, error: BailoError) {
    const outcome =
      error.code === 403
        ? {
            Description: 'User does not have permission to execute the request',
            Permitted: false,
            Success: false,
          }
        : {
            Description: error.message,
            Success: false,
          }
    const event = this.generateEvent(req, { url: req.originalUrl, httpMethod: req.method }, outcome)

    log.info(event, 'Logging Event')
  }

  generateEvent(req: Request, resourceInfo: object, outcome?: Outcome) {
    return {
      EventTime: new Date().toString(),
      EventSource: {
        System: 'BAILO',
        Device: {
          Hostname: 'http://localhost:8080',
        },
        Client: req.ip,
        User: req.user,
      },
      EventDetail: {
        typeId: req.audit.typeId,
        description: req.audit.description,
        resource: JSON.stringify(resourceInfo),
        ...(outcome && { outcome }),
      },
    }
  }

  onCreateSchema(req: Request, schema: SchemaInterface) {
    this.checkEventType(AuditInfo.CreateSchema, req)
    const event = this.generateEvent(req, { id: schema.id })
    log.info(event, 'Logging Event')
  }

  onViewSchema(req: Request, schema: SchemaInterface) {
    this.checkEventType(AuditInfo.ViewSchema, req)
    const event = this.generateEvent(req, { id: schema.id })
    log.info(event, 'Logging Event')
  }

  onSearchSchemas(req: Request, schemas: SchemaInterface[]) {
    this.checkEventType(AuditInfo.SearchSchemas, req)
    const event = this.generateEvent(req, {
      url: req.originalUrl,
      results: schemas.map((schema) => ({ id: schema.id })),
    })
    log.info(event, 'Logging Event')
  }

  onSearchImages(req: Request, images: { repository: string; name: string; tags: string[] }[]) {
    this.checkEventType(AuditInfo.SearchImages, req)
    const event = this.generateEvent(req, {
      url: req.originalUrl,
      results: images.map((image) => ({ repository: image.repository, name: image.name })),
    })
    log.info(event, 'Logging Event')
  }

  checkEventType(auditInfo: AuditInfoKeys, req: Request) {
    if (auditInfo.typeId !== req.audit.typeId && auditInfo.description !== req.audit.description) {
      throw new Error(`Audit: Expected type '${JSON.stringify(auditInfo)}' but recieved '${JSON.stringify(req.audit)}'`)
    }
  }
}
