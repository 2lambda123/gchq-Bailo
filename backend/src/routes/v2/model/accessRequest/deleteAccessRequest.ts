import bodyParser from 'body-parser'
import { Request, Response } from 'express'
import { z } from 'zod'

import { removeAccessRequest } from '../../../../services/v2/accessRequest.js'
import { parse } from '../../../../utils/validate.js'

export const deleteAccessRequestSchema = z.object({
  params: z.object({
    modelId: z.string(),
    accessRequestId: z.string(),
  }),
})

interface DeleteAccessRequestResponse {
  message: string
}

export const deleteAccessRequest = [
  bodyParser.json(),
  async (req: Request, res: Response<DeleteAccessRequestResponse>) => {
    const {
      params: { accessRequestId },
    } = parse(req, deleteAccessRequestSchema)

    await removeAccessRequest(req.user, accessRequestId)

    return res.json({
      message: 'Successfully removed access request.',
    })
  },
]
