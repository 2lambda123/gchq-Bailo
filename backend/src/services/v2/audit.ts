import { Request } from 'express'
import { IncomingMessage, ServerResponse } from 'http'

import log from './log.js'

interface Event {
  EventTime: string
  EventSource: {
    System: string
    Generator: string
    Device: {
      Hostname: string
    }
    Client: string
    User: string
  }
  EventDetail: EventDetail
}
export interface EventDetail {
  TypeId: string
  Description: string
}

const exemptions: string[] = ['u/r/l']

export async function audit(this: ServerResponse) {
  if (this.req.url && exemptions.includes(this.req.url)) {
    //Decide if the request needs auditing and whether it can be auditted
    log.info({ url: this.req.url }, 'Endpoint not required to be auditted.')
    return
  } else if (!isRequest(this.req)) {
    log.error(
      { url: this.req.url },
      'Could not audit. This should not happen and only endpoints in the exemptions list should fail to audit.',
    )
    return
  }

  //Create the event object
  const event = createEvent(this.req)

  // Process event depending on auditting system in config
  // Write to file, send to service, formatting etc
  log.info(event, 'Logging Event')
}

function createEvent(req: Request): Event {
  return {
    EventTime: new Date().toString(),
    EventSource: {
      System: 'BAILO',
      Generator: 'BAILO',
      Device: {
        Hostname: 'http://localhost:8080',
      },
      Client: req.ip,
      User: req.user,
    },
    EventDetail: req.eventDetail,
  }
}
export function isRequest(request: unknown): request is Request {
  if (typeof request !== 'object' || request === null) {
    return false
  }

  if (!(request['eventDetail'] && request['user'])) {
    return false
  }

  if (request instanceof IncomingMessage) {
    return true
  }

  return false
}
