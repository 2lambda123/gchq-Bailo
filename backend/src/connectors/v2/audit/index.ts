import config from '../../../utils/v2/config.js'
import { ConfigurationError } from '../../../utils/v2/error.js'
import { BaseAuditConnector } from './Base.js'
import { SillyAuditConnector } from './silly.js'
import { StdoutAuditConnector } from './stdout.js'

export const AuditKind = {
  Silly: 'silly',
  Stdout: 'stdout',
} as const
export type AuditKindKeys = (typeof AuditKind)[keyof typeof AuditKind]

let auditConnector: undefined | BaseAuditConnector = undefined
export function getAuditConnector(cache = true) {
  if (auditConnector && cache) {
    return auditConnector
  }

  switch (config.connectors.audit.kind) {
    case AuditKind.Silly:
      auditConnector = new SillyAuditConnector()
      break
    case AuditKind.Stdout:
      auditConnector = new StdoutAuditConnector()
      break
    default:
      throw ConfigurationError(`'${config.connectors.audit.kind}' is not a valid audit kind.`, {
        validKinds: Object.values(AuditKind),
      })
  }

  return auditConnector
}

export default getAuditConnector()
