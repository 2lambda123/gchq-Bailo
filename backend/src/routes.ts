import parser from 'body-parser'
import MongoStore from 'connect-mongo'
import express from 'express'
import session from 'express-session'
import grant from 'grant'

import { expressErrorHandler as expressErrorHandlerV2 } from './routes/middleware/expressErrorHandler.js'
import { expressLogger as expressLoggerV2 } from './routes/middleware/expressLogger.js'
import { getUser as getUserV2 } from './routes/middleware/getUser.js'
import { getApplicationLogs, getItemLogs } from './routes/v1/admin.js'
import { getApprovals, getNumApprovals, postApprovalResponse } from './routes/v1/approvals.js'
import {
  fetchRawModelFiles,
  getDeployment,
  getDeploymentAccess,
  getDeploymentApprovals,
  getExportModelVersion,
  getUserDeployments,
  postDeployment,
  postUngovernedDeployment,
  resetDeploymentApprovals,
} from './routes/v1/deployment.js'
import {
  getModelAccess,
  getModelById,
  getModelByUuid,
  getModelDeployments,
  getModels,
  getModelSchema,
  getModelVersion,
  getModelVersions,
} from './routes/v1/model.js'
import { getDockerRegistryAuth } from './routes/v1/registryAuth.js'
import { getDefaultSchema, getSchema, getSchemas, postSchema } from './routes/v1/schema.js'
import { getSpecification } from './routes/v1/specification.js'
import { getUiConfig } from './routes/v1/uiConfig.js'
import { postUpload } from './routes/v1/upload.js'
import { favouriteModel, getLoggedInUser, getUsers, postRegenerateToken, unfavouriteModel } from './routes/v1/users.js'
import {
  deleteVersion,
  getVersion,
  getVersionAccess,
  getVersionApprovals,
  getVersionFile,
  getVersionFileList,
  postRebuildModel,
  postResetVersionApprovals,
  putUpdateLastViewed,
  putVersion,
} from './routes/v1/version.js'
import { deleteAccessRequest } from './routes/v2/model/accessRequest/deleteAccessRequest.js'
import { getAccessRequest } from './routes/v2/model/accessRequest/getAccessRequest.js'
import { getModelAccessRequests } from './routes/v2/model/accessRequest/getModelAccessRequests.js'
import { patchAccessRequest } from './routes/v2/model/accessRequest/patchAccessRequest.js'
import { postAccessRequest } from './routes/v2/model/accessRequest/postAccessRequest.js'
import { deleteFile } from './routes/v2/model/file/deleteFile.js'
import { getDownloadFile } from './routes/v2/model/file/getDownloadFile.js'
import { getFiles } from './routes/v2/model/file/getFiles.js'
import { postFinishMultipartUpload } from './routes/v2/model/file/postFinishMultipartUpload.js'
import { postSimpleUpload } from './routes/v2/model/file/postSimpleUpload.js'
import { postStartMultipartUpload } from './routes/v2/model/file/postStartMultipartUpload.js'
import { getModel } from './routes/v2/model/getModel.js'
import { getModelsSearch } from './routes/v2/model/getModelsSearch.js'
import { getImages } from './routes/v2/model/images/getImages.js'
import { getModelCard } from './routes/v2/model/modelcard/getModelCard.js'
import { getModelCardRevisions } from './routes/v2/model/modelcard/getModelCardRevisions.js'
import { postFromSchema } from './routes/v2/model/modelcard/postFromSchema.js'
import { putModelCard } from './routes/v2/model/modelcard/putModelCard.js'
import { patchModel } from './routes/v2/model/patchModel.js'
import { postModel } from './routes/v2/model/postModel.js'
import { getModelCurrentUserRoles } from './routes/v2/model/roles/getModelCurrentUserRoles.js'
import { getModelRoles } from './routes/v2/model/roles/getModelRoles.js'
import { deleteRelease } from './routes/v2/release/deleteRelease.js'
import { getRelease } from './routes/v2/release/getRelease.js'
import { getReleases } from './routes/v2/release/getReleases.js'
import { postRelease } from './routes/v2/release/postRelease.js'
import { getReviews } from './routes/v2/review/getReviews.js'
import { postAccessRequestReviewResponse } from './routes/v2/review/postAccessRequestReviewResponse.js'
import { postReleaseReviewResponse } from './routes/v2/review/postReleaseReviewResponse.js'
import { getSchema as getSchemaV2 } from './routes/v2/schema/getSchema.js'
import { getSchemas as getSchemasV2 } from './routes/v2/schema/getSchemas.js'
import { postSchema as postSchemaV2 } from './routes/v2/schema/postSchema.js'
import { patchTeam } from './routes/v2/team/getMyTeams.js'
import { getTeam } from './routes/v2/team/getTeam.js'
import { getTeams } from './routes/v2/team/getTeams.js'
import { postTeam } from './routes/v2/team/postTeam.js'
import { getCurrentUser } from './routes/v2/user/getCurrentUser.js'
import config from './utils/config.js'
import logger, { expressErrorHandler, expressLogger } from './utils/logger.js'
import { getUser } from './utils/user.js'

export const server = express()

if (config.oauth.enabled) {
  server.use(
    session({
      secret: config.session.secret,
      resave: true,
      saveUninitialized: true,
      cookie: { maxAge: 30 * 24 * 60 * 60000 }, // store for 30 days
      store: MongoStore.create({
        mongoUrl: config.mongo.uri,
      }),
    }),
  )
}

server.use('/api/v1', getUser)
server.use('/api/v1', expressLogger)

if (config.experimental.v2) {
  server.use('/api/v2', getUserV2)
  server.use('/api/v2', expressLoggerV2)
}

if (config.oauth.enabled) {
  server.use(parser.urlencoded({ extended: true }))
  server.use(grant.default.express(config.oauth.grant))

  server.get('/api/login', (req, res) => {
    res.redirect(`/api/connect/${config.oauth.provider}/login`)
  })

  server.get('/api/logout', (req, res) => {
    req.session.destroy(function (err: unknown) {
      if (err) throw err
      res.redirect('/')
    })
  })
}

// V1 APIs

server.post('/api/v1/model', ...postUpload)

server.get('/api/v1/models', ...getModels)
server.get('/api/v1/model/uuid/:uuid', ...getModelByUuid)
server.get('/api/v1/model/id/:id', ...getModelById)
server.get('/api/v1/model/:uuid/schema', ...getModelSchema)
server.get('/api/v1/model/:uuid/versions', ...getModelVersions)
server.get('/api/v1/model/:uuid/version/:version', ...getModelVersion)
server.get('/api/v1/model/:uuid/deployments', ...getModelDeployments)
server.get('/api/v1/model/:uuid/access', ...getModelAccess)

server.post('/api/v1/deployment', ...postDeployment)
server.post('/api/v1/deployment/ungoverned', ...postUngovernedDeployment)
server.get('/api/v1/deployment/:uuid', ...getDeployment)
server.get('/api/v1/deployment/:id/approvals', ...getDeploymentApprovals)
server.get('/api/v1/deployment/user/:id', ...getUserDeployments)
server.post('/api/v1/deployment/:uuid/reset-approvals', ...resetDeploymentApprovals)
server.get('/api/v1/deployment/:uuid/access', ...getDeploymentAccess)
server.get('/api/v1/deployment/:uuid/version/:version/raw/:fileType', ...fetchRawModelFiles)
server.get('/api/v1/deployment/:uuid/version/:version/export', ...getExportModelVersion)

server.get('/api/v1/version/:id', ...getVersion)
server.get('/api/v1/version/:id/contents/:file/list', ...getVersionFileList)
server.get('/api/v1/version/:id/contents/:file', ...getVersionFile)
server.get('/api/v1/version/:id/access', ...getVersionAccess)
server.get('/api/v1/version/:id/approvals', ...getVersionApprovals)
server.put('/api/v1/version/:id', ...putVersion)
server.delete('/api/v1/version/:id', ...deleteVersion)
server.post('/api/v1/version/:id/reset-approvals', ...postResetVersionApprovals)
server.post('/api/v1/version/:id/rebuild', ...postRebuildModel)
server.put('/api/v1/version/:id/lastViewed/:role', ...putUpdateLastViewed)

server.get('/api/v1/schemas', ...getSchemas)
server.get('/api/v1/schema/default', ...getDefaultSchema)
server.get('/api/v1/schema/:ref', ...getSchema)
server.post('/api/v1/schema', ...postSchema)

server.get('/api/v1/config/ui', ...getUiConfig)

server.get('/api/v1/users', ...getUsers)
server.get('/api/v1/user', ...getLoggedInUser)
server.post('/api/v1/user/token', ...postRegenerateToken)
server.post('/api/v1/user/favourite/:id', ...favouriteModel)
server.post('/api/v1/user/unfavourite/:id', ...unfavouriteModel)

server.get('/api/v1/approvals', ...getApprovals)
server.get('/api/v1/approvals/count', ...getNumApprovals)
server.post('/api/v1/approval/:id/respond', ...postApprovalResponse)

server.get('/api/v1/registry_auth', ...getDockerRegistryAuth)

server.get('/api/v1/specification', ...getSpecification)

server.get('/api/v1/admin/logs', ...getApplicationLogs)
server.get('/api/v1/admin/logs/build/:buildId', ...getItemLogs)
server.get('/api/v1/admin/logs/approval/:approvalId', ...getItemLogs)

/**
 ** V2 API **
 */

if (config.experimental.v2) {
  logger.info('Using experimental V2 endpoints')

  server.post('/api/v2/models', ...postModel)
  server.get('/api/v2/models/search', ...getModelsSearch)
  // server.post('/api/v2/models/import', ...postModelImport)

  server.get('/api/v2/model/:modelId', ...getModel)
  server.patch('/api/v2/model/:modelId', ...patchModel)

  server.get('/api/v2/model/:modelId/model-card/:version', ...getModelCard)
  server.get('/api/v2/model/:modelId/model-card-revisions', ...getModelCardRevisions)
  server.put('/api/v2/model/:modelId/model-cards', ...putModelCard)

  // *server.get('/api/v2/template/models', ...getModelTemplates)
  // *server.post('/api/v2/model/:modelId/setup/from-template', ...postFromTemplate)
  // *server.post('/api/v2/model/:modelId/setup/from-existing', ...postFromExisting)
  server.post('/api/v2/model/:modelId/setup/from-schema', ...postFromSchema)

  server.post('/api/v2/model/:modelId/releases', ...postRelease)
  server.get('/api/v2/model/:modelId/releases', ...getReleases)
  server.get('/api/v2/model/:modelId/release/:semver', ...getRelease)
  server.delete('/api/v2/model/:modelId/release/:semver', ...deleteRelease)
  server.post('/api/v2/model/:modelId/release/:semver/review', ...postReleaseReviewResponse)

  server.post('/api/v2/model/:modelId/access-requests', ...postAccessRequest)
  server.get('/api/v2/model/:modelId/access-requests', getModelAccessRequests)
  server.get('/api/v2/model/:modelId/access-request/:accessRequestId', ...getAccessRequest)
  server.delete('/api/v2/model/:modelId/access-request/:accessRequestId', ...deleteAccessRequest)
  server.patch('/api/v2/model/:modelId/access-request/:accessRequestId', ...patchAccessRequest)
  server.post('/api/v2/model/:modelId/access-request/:accessRequestId/review', ...postAccessRequestReviewResponse)

  server.get('/api/v2/model/:modelId/files', ...getFiles)
  server.get('/api/v2/model/:modelId/file/:fileId/download', ...getDownloadFile)
  server.post('/api/v2/model/:modelId/files/upload/simple', ...postSimpleUpload)
  server.post('/api/v2/model/:modelId/files/upload/multipart/start', ...postStartMultipartUpload)
  server.post('/api/v2/model/:modelId/files/upload/multipart/finish', ...postFinishMultipartUpload)
  server.delete('/api/v2/model/:modelId/file/:fileId', ...deleteFile)

  server.get('/api/v2/model/:modelId/images', ...getImages)
  // *server.delete('/api/v2/model/:modelId/images/:imageId', ...deleteImage)

  // *server.get('/api/v2/model/:modelId/release/:semver/file/:fileCode/list', ...getModelFileList)
  // *server.get('/api/v2/model/:modelId/release/:semver/file/:fileCode/raw', ...getModelFileRaw)

  server.get('/api/v2/schemas', ...getSchemasV2)
  server.get('/api/v2/schema/:schemaId', ...getSchemaV2)
  server.post('/api/v2/schemas', ...postSchemaV2)

  server.get('/api/v2/reviews', ...getReviews)

  server.get('/api/v2/model/:modelId/roles', ...getModelRoles)
  server.get('/api/v2/model/:modelId/roles/mine', ...getModelCurrentUserRoles)

  server.post('/api/v2/teams', ...postTeam)
  server.get('/api/v2/teams', ...getTeams)
  server.get('/api/v2/teams/mine', ...getTeams)

  server.get('/api/v2/team/:teamId', ...getTeam)
  server.patch('/api/v2/team/:teamId', ...patchTeam)

  // server.post('/api/v2/teams/:teamId/members', ...postTeamMember)
  // server.get('/api/v2/teams/:teamId/members', ...getTeamMembers)
  // server.delete('/api/v2/teams/:teamId/members/:memberId', ...deleteTeamMember)
  // server.patch('/api/v2/teams/:teamId/members/:memberId', ...patchTeamMember)

  // server.get('/api/v2/teams/:teamId/roles/:memberId', ...getTeamMemberRoles)

  // server.get('/api/v2/users', ...getUsers)
  server.get('/api/v2/users/me', ...getCurrentUser)

  // server.post('/api/v2/user/:userId/tokens', ...postUserToken)
  // server.get('/api/v2/user/:userId/tokens', ...getUserTokens)
  // server.get('/api/v2/user/:userId/token/:tokenId', ...getUserToken)
  // server.delete('/api/v2/user/:userId/token/:tokenId', ...deleteUserToken)
} else {
  logger.info('Not using experimental V2 endpoints')
}

server.use('/api/v1', expressErrorHandler)

if (config.experimental.v2) {
  server.use('/api/v2', expressErrorHandlerV2)
}
