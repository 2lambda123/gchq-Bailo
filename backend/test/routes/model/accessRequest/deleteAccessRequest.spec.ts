import { describe, expect, test, vi } from 'vitest'

import audit from '../../../../src/connectors/v2/audit/__mocks__/index.js'
import { deleteAccessRequestSchema } from '../../../../src/routes/v2/model/accessRequest/deleteAccessRequest.js'
import { createFixture, testDelete } from '../../../testUtils/routes.js'

vi.mock('../../../../src/utils/user.js')
vi.mock('../../../../src/utils/v2/config.js')
vi.mock('../../../../src/connectors/v2/audit/index.js')
vi.mock('../../../../src/connectors/v2/authorisation/index.js')

describe('routes > accessRequest > deleteAccessRequest', () => {
  test('200 > ok', async () => {
    vi.mock('../../../../src/services/v2/accessRequest.js', () => ({
      removeAccessRequest: vi.fn(() => ({ message: 'Successfully removed access request.' })),
    }))

    const fixture = createFixture(deleteAccessRequestSchema)
    const res = await testDelete(
      `/api/v2/model/${fixture.params.modelId}/access-request/${fixture.params.accessRequestId}`,
    )

    expect(res.statusCode).toBe(200)
    expect(res.body).matchSnapshot()
  })

  test('audit > expected call', async () => {
    vi.mock('../../../../src/services/v2/accessRequest.js', () => ({
      removeAccessRequest: vi.fn(() => ({ message: 'Successfully removed access request.' })),
    }))

    const fixture = createFixture(deleteAccessRequestSchema)
    const res = await testDelete(
      `/api/v2/model/${fixture.params.modelId}/access-request/${fixture.params.accessRequestId}`,
    )

    expect(res.statusCode).toBe(200)
    expect(audit.onDeleteAccessRequest).toBeCalled()
    expect(audit.onDeleteAccessRequest.mock.calls.at(0).at(1)).toMatchSnapshot()
  })
})
