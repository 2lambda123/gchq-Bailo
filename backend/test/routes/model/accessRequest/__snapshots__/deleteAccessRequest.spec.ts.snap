// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`routes > accessRequest > deleteAccessRequest > 200 > ok 1`] = `
{
  "message": "Successfully removed access request.",
}
`;
